core/nightshade/src/nightshade.rs

```diff
@@ -0,0 +1,781 @@
+#[derive(Debug, Clone, Serialize)]
+pub struct State {
+    /// Triplet that describe the state
+    pub bare_state: BareState,
+    /// Proof for `primary_confidence`.
+    primary_proof: Option<Proof>,
+    /// Proof for `secondary_bare_state`.
+    secondary_proof: Option<Proof>,
+    /// Signature of the authority emitting this state
+    #[serde(with = "bs58_serializer")]
+    signature: BlsSignature,
+}
+
+macro_rules! check_true {
+    ($condition:expr) => {
+        if !$condition{
+            return false;
+        }
+    };
+}
+
+impl State {
+    /// Create new state
+    fn new(author: AuthorityId, hash: CryptoHash, secret_key: &BlsSecretKey) -> Self {
+        let bare_state = BareState::new(author, hash);
+        let signature = bare_state.sign(&secret_key);
+        Self {
+            bare_state,
+            primary_proof: None,
+            secondary_proof: None,
+            signature,
+        }
+    }
+
+    /// Create state with empty triplet.
+    /// See `BareState::empty` for more information
+    ///
+    /// Note: The signature of this state is going to be incorrect, but this state
+    /// will never be sended to other participants as current state.
+    fn empty() -> Self {
+        Self {
+            bare_state: BareState::empty(),
+            primary_proof: None,
+            secondary_proof: None,
+            signature: BlsSignature::empty(),
+        }
+    }
+
+    /// Create new State with increased confidence using `proof`
+    fn increase_confidence(&self, proof: Proof, secret_key: &BlsSecretKey) -> Self {
+        let bare_state = BareState {
+            primary_confidence: self.bare_state.primary_confidence + 1,
+            endorses: self.bare_state.endorses.clone(),
+            secondary_confidence: self.bare_state.secondary_confidence,
+        };
+
+        let signature = bare_state.sign(&secret_key);
+
+        Self {
+            bare_state,
+            primary_proof: Some(proof),
+            secondary_proof: self.secondary_proof.clone(),
+            signature,
+        }
+    }
+
+    /// Returns whether an authority having this triplet should commit to this triplet outcome.
+    fn can_commit(&self) -> bool {
+        self.bare_state.primary_confidence >= self.bare_state.secondary_confidence + COMMIT_THRESHOLD
+    }
+
+    /// Check if this state has correct proofs about the triplet it contains.
+    /// Each authority will check if this state is valid only if it has not successfully verified another
+    /// state with the same triplet before. Once it has verified that at least one authority has such
+    /// triplet, it accepts all further states with the same triplet.
+    ///
+    /// # Arguments
+    ///
+    /// * `authority` - The authority that send this state.
+    /// * `public_keys` - Public key of every authority in the network
+    fn verify(&self, authority: AuthorityId, public_keys: &Vec<BlsPublicKey>) -> bool {
```

Comment:
In the next PR, I suggest returning `Result<(), NSVerifyErr>`, where `NSVerifyErr` is an enum that enumerates various verification failure scenarios (with possibly added string descriptions with detail for errors that are applicable). Then `check_true!` will be replaced, e.g. `check_true!(self.bare_state.verify());` will turn into `self.bare_state.verify()?;`

Reply:
Returning `Result<(), NSVerifyErr>` is a great idea. I'll write/explain all types of failure on Nomicon, it will help to catch errors easily. I replace macro with `check_or!`, it is still useful for plain boolean checks.